### PR TITLE
(PDB-1674) PuppetDB hangs on shutdown when the database is down 

### PR DESCRIPTION
**Pull request by senior**
This adds connection timeouts to the command processing and query
connection pools. This fixes the problem where the ActiveMQ session
can't be properly shutdown because the command connection pool can't get
a connection to the database (i.e. the database is down). Previously
this would hang, even with a Control-C. This hang was due to the
connection pool software getting caught in a loop trying to get a
connection to the database.

This connection timeout was also applied to the query side of the
application as user provided queries are expected to respond quickly and
a hanging behavior would be bad there as well.

This blocking/retrying behavior is what we want when PuppetDB first
starts up as we need to initialize the schema etc. This commit also
creates a special connection pool (just for startup) that doesn't have
these timeouts. This pool is closed after the startup code that needs a
database connection is finished.

### CODE REVIEW

**Comment by senior** [comment]
This just needs docs for the new connection-timeout config

**Comment by senior** [comment]
@pljenkinsro retest this please

File: src/puppetlabs/puppetdb/cli/services.clj [hunk 1]
```diff
@@ -237,6 +237,35 @@
...
+(defn init-with-db
+  "All initialization operations needing a database connection should
+  happen here. This function creates a connection pool using
+  `write-db-config` that will hang until it is able to make a
+  connection to the database. This covers the case of the database not
+  being fully started when PuppetDB starts. This connection pool will
+  be opened and closed within the body of this function."
+  [write-db-config product-name]
+  (with-open [init-db-pool (pl-jdbc/make-connection-pool (assoc write-db-config
+                                                           ;; Block waiting to grab a connection
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/cli/services.clj]: this alignment looks off

Comment (senior) [inline src/puppetlabs/puppetdb/cli/services.clj]: Maybe the comment is what throws it off? I reformatted it and nothing changed.

**Comment by senior** [comment]
Added some docs, switched from 2ms connection timeout on the query side to 500ms. Travis is failing with connection timeouts, only on PG. I'm wondering if the issue it's having is lazy init. The first poll for a database connection needs to init the pool, which maybe takes more than 2ms?

**Comment by senior** [comment]
1 transient failing cell ^^^

**Comment by rbrw** [comment]
Not sure which test is failing, but I also wouldn't be surprised if some tests might need to be adjusted to account for these changes, i.e. a query EAGAIN is now more "normal".

File: documentation/configure.markdown [hunk 2]
```diff
@@ -370,6 +370,10 @@ The maximum time (in minutes) a pooled connection should remain open. Any connec
 
 If not supplied, we won't terminate connections based on their age alone.
 
+### `connection-timeout`
+
+The maximum time to wait (in milliseconds) to acquire a connection from the pool of database connections. If not supplied, defaults to 1000.
+
 ###`statements-cache-size`
```
Comment (rbrw) [inline documentation/configure.markdown]: Perhaps wrap this (and below)?

File: src/puppetlabs/puppetdb/cli/services.clj [hunk 3]
```diff
@@ -237,6 +237,35 @@
...
 
+(defn initialize-schema
+  "Ensure the database is migrated to the latest version, and warn if
+  it's deprecated, log and exit if it's unsupported. We do this in a
+  single connection because HSQLDB seems to get confused if the
+  database doesn't exist but we open and close a connection without
+  creating anything."
+  [db-conn-pool product-name]
+  (sql/with-connection db-conn-pool
+    (scf-store/validate-database-version #(System/exit 1))
```
Comment (rbrw) [inline src/puppetlabs/puppetdb/cli/services.clj]: Be glad when we finally get a chance to consider our System/exits (but not an issue for this PR).

**Comment by rbrw** [comment]
This all looks good to me, works fine, and I liked the commit breakdown too.

File: documentation/configure.markdown (hunk 2 above)
Comment (senior) [inline documentation/configure.markdown]: Meh, I do like the wrap there, but the stuff around it is not wrapped. The whole file goes in and out of it, but specifically those sections are not. Probably something we should just do on all files at once if we want the change.

**Pull request merged**